### PR TITLE
Reduce getting node api calls when updating node status

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

When kubelet updating node status, firstly get node from local cache instead of get from apiserver. 
This will reduce get node api call, thus reduce apiserver burden, especially in large scale cluster. 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
/assign @derekwaynecarr 
cc @liggitt @wojtek-t

File: pkg/kubelet/kubelet_node_status.go [hunk 1]
```diff
@@ -370,13 +369,18 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
...
-	opts := metav1.GetOptions{}
+	var node *v1.Node
+	var err error
 	if tryNumber == 0 {
-		util.FromApiserverCache(&opts)
-	}
-	node, err := kl.heartbeatClient.Nodes().Get(string(kl.nodeName), opts)
-	if err != nil {
-		return fmt.Errorf("error getting node %q: %v", kl.nodeName, err)
+		node, err = kl.nodeInfo.GetNodeInfo(string(kl.nodeName))
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: How does GetNodeInfo maintain its current view of the node object?

This really doesn't seem safe. We're computing a patch to send to the server to modify server state based on a local view of the current state. This seems likely to result in patches that are accepted/applied by the server but do not end up with the desired node object persisted. 

As an example, say the apiserver has a condition on the node that doesn't match the kubelet's view of the condition. The kubelet will not include the condition change in the outgoing patch, since from its perspective, nothing has changed in the node.

Comment (wojtek-t) [reply pkg/kubelet/kubelet_node_status.go]: Technically, even without this change, we were getting the node from etcd directly, but from apiserver cache (that is propagated via watch and can be outdated).

IIRC, the nodeInfo is also propagated via watch. So yes - this may be even more delayed. The question is how worse that situation is.
In general, the scenario you described might have been true even before - it's just a bit more probably after that change. Both before and after this change, this is eventually consistent.

Personally, I don't have strong preference either way. I would say this change is fine, but I also see your concerns.

Comment (liggitt) [reply pkg/kubelet/kubelet_node_status.go]: a stale/stuck watch from in the kubelet could result in the situation I described indefinitely. 

can we tell if the watch-cache-fed version has at least observed the last successful heartbeat? That would indicate it is relatively up to date

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: Even observed the last successful heartbeat(or even directly get from etcd), it can not prevent `This seems likely to result in patches that are accepted/applied by the server but do not end up with the desired node object persisted.`  Because Get and Patch are not atomic. 

But as @wojtek-t said `Both before and after this change, this is eventually consistent.`

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: Yeah, that can be protected by `compare resource version`

Comment (liggitt) [reply pkg/kubelet/kubelet_node_status.go]: > that can be protected by compare resource version

no comparisons that treat it as a non-opaque value. Checking the heartbeat timestamp to see if it matches the last one we set would be ok (would let us know we observed our last update)

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: resonable

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: Done

File: pkg/kubelet/kubelet_node_status.go [hunk 2]
```diff
@@ -57,6 +56,9 @@ const (
 	maxNamesPerImageInNodeStatus = 5
 )
 
+// Last time we patch node status.
+var lastHeartbeatTime = metav1.NewTime(time.Now())
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: don't use a package var for this... if we need to track it, do so in the Kubelet struct, and ensure access to it is threadsafe

File: pkg/kubelet/kubelet_node_status.go [hunk 3]
```diff
@@ -370,13 +372,21 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
...
+	var err error
 	if tryNumber == 0 {
-		util.FromApiserverCache(&opts)
-	}
-	node, err := kl.heartbeatClient.Nodes().Get(string(kl.nodeName), opts)
-	if err != nil {
-		return fmt.Errorf("error getting node %q: %v", kl.nodeName, err)
+		node, err = kl.nodeInfo.GetNodeInfo(string(kl.nodeName))
+		if err != nil {
+			return fmt.Errorf("error getting node from cache %q: %v", kl.nodeName, err)
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: an error fetching from cache should fall back to the current path, not error

File: pkg/kubelet/kubelet_node_status.go [hunk 4]
```diff
@@ -370,13 +372,21 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
...
+		}
+		if len(node.Status.Conditions) > 0 && node.Status.Conditions[0].LastHeartbeatTime.Before(&lastHeartbeatTime) {
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: don't assume the first condition in the list is one set by the node with a heartbeat time... look for a specific condition we know the kublet sets

Comment (liggitt) [reply pkg/kubelet/kubelet_node_status.go]: @wojtek-t how does this interact with your proposal to move heartbeat timestamps out of the node status?

File: pkg/kubelet/kubelet_node_status.go [hunk 5]
```diff
@@ -394,6 +404,10 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
 	if err != nil {
 		return err
 	}
+	if len(updatedNode.Status.Conditions) > 0 {
+		lastHeartbeatTime = updatedNode.Status.Conditions[0].LastHeartbeatTime
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: same here, don't grab the timestamp of the first condition. an alternative could be to grab the latest timestamp of all the conditions that provide a non-zero heartbeat timestamp

**Comment by derekwaynecarr** [comment]
@hzxuzhonghu can you quantify how much this reduces traffic?  for x number of nodes, what is the old and new measurement you see w/ this change?

File: pkg/kubelet/kubelet_node_status.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: We have `nodeStatusUpdateRetry = 5` retry on higher loop.

File: pkg/kubelet/kubelet_node_status.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: ok

File: pkg/kubelet/kubelet_node_status.go (hunk 5 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: ok

**Comment by hzxuzhonghu** [comment]
@derekwaynecarr In theory, in 5000 nodes cluster we can reduce 5000/10 qps.

File: pkg/kubelet/kubelet_node_status.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: No multi-thread access now. move to  Kubelet struct sounds ok.

File: pkg/kubelet/kubelet_node_status.go (hunk 3 above)
Comment (liggitt) [reply pkg/kubelet/kubelet_node_status.go]: No, don't waste an attempt on a faulty cache, fall back to a live get immediately

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: ic

**Comment by hzxuzhonghu** [comment]
@derekwaynecarr @liggitt for approve

File: pkg/kubelet/kubelet_node_status.go [hunk 6]
```diff
@@ -370,13 +369,26 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
 	// apiserver cache (the data might be slightly delayed but it doesn't
 	// seem to cause more conflict - the delays are pretty small).
 	// If it result in a conflict, all retries are served directly from etcd.
-	opts := metav1.GetOptions{}
+	var node *v1.Node
+	var err error
 	if tryNumber == 0 {
-		util.FromApiserverCache(&opts)
+		node, err = kl.nodeInfo.GetNodeInfo(string(kl.nodeName))
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: rather than set `node` and then clear, prefer setting a local var and then assigning to `node` when heartbeat matches:

```
cachedNode, err := kl.nodeInfo.GetNodeInfo(string(kl.nodeName))
switch {
case err != nil:
  glog.Warningf("error getting node from cache %q: %v", kl.nodeName, err)
case !kl.getNodeReadyLastHeartbeatTime(cachedNode).Equal(kl.lastHeartbeatTime):
  glog.Infof("cached node heartbeat time does not match last submitted time")
default:
  node = cachedNode
}
```

File: pkg/kubelet/kubelet_test.go [hunk 7]
```diff
@@ -158,6 +158,7 @@ func newTestKubeletWithImageList(
 	kubelet.recorder = fakeRecorder
 	kubelet.kubeClient = fakeKubeClient
 	kubelet.heartbeatClient = fakeKubeClient.CoreV1()
+	kubelet.lastHeartbeatTime = metav1.Now()
```
Comment (liggitt) [inline pkg/kubelet/kubelet_test.go]: why was this change needed?

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_test.go]: just fix test failure

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_test.go]: getNodeReadyLastHeartbeatTime(cachedNode) returns unset time in this test case.

Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_test.go]: @liggitt pushed an alternative way to fix ut, ptal.

File: pkg/kubelet/kubelet_node_status.go [hunk 8]
```diff
@@ -1100,6 +1116,18 @@ func (kl *Kubelet) defaultNodeStatusFuncs() []func(*v1.Node) error {
...
+// getNodeReadyLastHeartbeatTime get the LastHeartbeatTime on NodeReady condition.
+func (kl *Kubelet) getNodeReadyLastHeartbeatTime(node *v1.Node) *metav1.Time {
+	var lastHeartbeatTime *metav1.Time
+	for i := range node.Status.Conditions {
+		if node.Status.Conditions[i].Type == v1.NodeReady {
+			lastHeartbeatTime = &node.Status.Conditions[i].LastHeartbeatTime
+			break
+		}
+	}
+	return lastHeartbeatTime
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: `return nil`

File: pkg/kubelet/kubelet_node_status.go [hunk 9]
```diff
@@ -1100,6 +1116,18 @@ func (kl *Kubelet) defaultNodeStatusFuncs() []func(*v1.Node) error {
 	}
 }
 
+// getNodeReadyLastHeartbeatTime get the LastHeartbeatTime on NodeReady condition.
+func (kl *Kubelet) getNodeReadyLastHeartbeatTime(node *v1.Node) *metav1.Time {
+	var lastHeartbeatTime *metav1.Time
+	for i := range node.Status.Conditions {
+		if node.Status.Conditions[i].Type == v1.NodeReady {
+			lastHeartbeatTime = &node.Status.Conditions[i].LastHeartbeatTime
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: `return &node.Status.Conditions[i].LastHeartbeatTime`

File: pkg/kubelet/kubelet_node_status.go [hunk 10]
```diff
@@ -1100,6 +1116,18 @@ func (kl *Kubelet) defaultNodeStatusFuncs() []func(*v1.Node) error {
 	}
 }
 
+// getNodeReadyLastHeartbeatTime get the LastHeartbeatTime on NodeReady condition.
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: doc it returns nil if the condition doesn't exist

File: pkg/kubelet/kubelet_node_status.go [hunk 11]
```diff
@@ -394,6 +405,11 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
 	if err != nil {
 		return err
 	}
+
+	if kl.getNodeReadyLastHeartbeatTime(updatedNode) != nil {
```
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: what about:

```go
if lastHeartbeat := kl.getNodeReadyLastHeartbeatTime(updatedNode); lastHeartbeat != nil {
  kl.lastHeartbeatTime = *lastHeartbeat
}
```

Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: @wojtek-t this bakes in assumptions about what `kl.setNodeStatus` does (that it updates the heartbeat timestamp of the node object). we know those assumptions are desired to change in the future (xref https://github.com/kubernetes/community/pull/2090). is @kubernetes/sig-node-pr-reviews ok with entangling this function somewhat in the short-term to get a performance gain at the cost of having to untangle it when node heartbeats get moved to a separate function?

**Comment by liggitt** [comment]
I'm tentatively ok with this PR pending sig-node approval (especially regarding https://github.com/kubernetes/kubernetes/pull/64065#discussion_r191628587)

@wojtek-t are there any scalability suites you want to run this through prior to merge?

File: pkg/kubelet/kubelet_node_status.go (hunk 11 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: ok

File: pkg/kubelet/kubelet_node_status.go (hunk 10 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: ok

File: pkg/kubelet/kubelet_node_status.go (hunk 9 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/kubelet_node_status.go]: ok

File: pkg/kubelet/kubelet_node_status.go (hunk 11 above)
Comment (wojtek-t) [reply pkg/kubelet/kubelet_node_status.go]: I'm not 100% sure that we will have to change that when implementing that proposal. But my personal feeling is that even if we have too, this change is small enough that it's worth it. WDYT?

**Comment by wojtek-t** [comment]
> @wojtek-t are there any scalability suites you want to run this through prior to merge?

I think the risk of this PR is very small. I took a look into presubmit results from it (kubemark-500 and they look ok. So I think it's fine to let it go as is.

**Comment by hzxuzhonghu** [comment]
@Random-Liu @dchen1107 @derekwaynecarr @tallclair  for approval

**Comment by hzxuzhonghu** [comment]
kindly ping @wojtek-t for approval

**Comment by wojtek-t** [comment]
Someone from node team should approve it.

File: pkg/kubelet/kubelet_node_status.go (hunk 11 above)
Comment (yujuhong) [reply pkg/kubelet/kubelet_node_status.go]: /cc @mtaufen, who's working on the heartbeat proposal.
